uimaj-core/src/test/java/org/apache/uima/cas/test/JCasClassLoaderTest.java

```diff
@@ -0,0 +1,292 @@
+          throws ResourceInitializationException, CASException, InvalidXMLException, IOException {
+    TypeSystemDescription tsd = getXMLParser().parseTypeSystemDescription(new XMLInputSource(
+            new File("src/test/resources/CASTests/desc/TokensAndSentencesTS.xml")));
+    ResourceManager resMgr = newDefaultResourceManager();
+    // resMgr.setExtensionClassPath(cl, "", false);
+    CASImpl cas = (CASImpl) createCas(tsd, null, null, null, resMgr);
+    cas.setJCasClassLoader(cl);
+    return cas.getJCas();
+  }
+
+  /**
+   * Creates a new analysis engine from the given class using the given classloader and sets it up
+   * so it used the given classloader to load its JCas wrappers.
+   */
+  private AnalysisEngine makeAnalysisEngine(Class<? extends Annotator_ImplBase> aeClass,
+          ClassLoader cl) throws ResourceInitializationException, MalformedURLException {
+    ResourceManager resMgr = newDefaultResourceManager();
+    resMgr.setExtensionClassPath(cl, "", false);
+    AnalysisEngineDescription desc = getResourceSpecifierFactory()
+            .createAnalysisEngineDescription();
+    desc.setAnnotatorImplementationName(aeClass.getName());
+    desc.setPrimitive(true);
+    return produceAnalysisEngine(desc, resMgr, null);
+  }
+
+  public static class AddATokenAnnotator extends JCasAnnotator_ImplBase {
+    @Override
+    public void process(JCas aJCas) throws AnalysisEngineProcessException {
+      System.out.printf("%s class loader: %s%n", getClass().getName(), getClass().getClassLoader());
+      
+      new Token(aJCas, 0, aJCas.getDocumentText().length()).addToIndexes();
+    }
+  }
+
+  public static class FetchTheTokenAnnotator extends JCasAnnotator_ImplBase {
+    @Override
+    public void process(JCas aJCas) throws AnalysisEngineProcessException {
+      System.out.printf("%s class loader: %s%n", getClass().getName(), getClass().getClassLoader());
+
+      List<Token> tokens = new ArrayList<>();
+      aJCas.getAllIndexedFS(Token.class).forEachRemaining(tokens::add);
+      
+      assertThat(tokens).isNotEmpty();
+    }
+  }
+
+  /**
+   * Special ClassLoader that helps us modeling different class loader topologies.
+   */
+  public static class IsolatingClassloader extends ClassLoader {
+
+    private final Set<String> hideClassesPatterns = new HashSet<>();
+    private final Set<String> redefineClassesPatterns = new HashSet<>();
+    private final Map<String, ClassLoader> delegates = new LinkedHashMap<>();
+    private final String id;
+
+    private Map<String, Class<?>> loadedClasses = new HashMap<>();
+
+    public IsolatingClassloader(String name, ClassLoader parent) {
+      super(parent);
+
+      id = name;
+    }
+    
+    public IsolatingClassloader hiding(String... patterns)
+    {
+      hideClassesPatterns.addAll(asList(patterns));
+      return this;
+    }
+
+    public IsolatingClassloader redefining(String... patterns)
+    {
+      hideClassesPatterns.addAll(asList(patterns));
```

Comment:
should this be adding these patterns to "redefineClassesPatterns", not "hideClassesPatterns"?

Reply:
Indes

Reply:
OK, see long comment in the Jira.  I'm thinking the test case has to be rewritten a bit?